tap_harvest/streams.py

```diff
@@ -0,0 +1,656 @@
+        if self.object_to_id is not None:
+            # Loop through each key of the object_to_id list.
+            for key in self.object_to_id:
+                key_object = row.get(key)
+                if key_object:
+                    row[key + '_id'] = key_object.get('id')
+                else:
+                    row[key + '_id'] = None
+
+    def sync_endpoint(self,
+                      client,
+                      catalog,
+                      config,
+                      state,
+                      tap_state,
+                      selected_streams,
+                      parent_row=None):
+        """
+        A common function to sync incremental streams.
+        """
+        # Retrieve schema and metadata of stream from the catalog
+        schema, stream_metadata = self.get_schema_and_metadata(catalog)
+        parent_row = parent_row or {}
+
+        bookmark_field = next(iter(self.replication_keys)) if self.replication_keys else None
+        children = self.children
+
+        current_time = datetime.now().strftime(DATE_FORMAT)
+        min_bookmark_among_parent_child = self.get_min_bookmark(self.tap_stream_id,
+                                        selected_streams,
+                                        current_time,
+                                        config['start_date'],
+                                        state)
+
+        # Get the latest bookmark for the stream and set the last_datetime
+        last_datetime = get_bookmark(self.tap_stream_id, state, config['start_date'])
+
+        with Transformer() as transformer:
+            page = 1
+            # Loop until the last page.
+            while page:
+
+                # Add parent_id in the URL to get records of child stream.
+                url = get_url(self.endpoint or self.tap_stream_id).format(
+                    parent_row.get(self.parent_id))
+
+                params = {}
+                if self.with_updated_since:
+                    params = {"updated_since": min_bookmark_among_parent_child}
+                params['page'] =  page
+
+                # Call API to fetch the records
+                response = client.request(url, params)
+                path = self.path or self.tap_stream_id
+                data = response[path]
+                time_extracted = utils.now()
+                for row in data:
+                    # Add fields at 1st level explicitly
+                    row = self.add_field_at_1st_level(row=row)
+                    self.convert_object_to_id(row)
+
+                    if self.parent_id:
+                        # Remove the last character `s` from the parent stream name and
+                        # join it with `_id` to save the parent id in the child record.
+                        # For example if a parent is `invoices`, then save parent id in
+                        # invoice_id key to the child.
+                        row[self.parent[:-1]+'_id'] = parent_row.get(self.parent_id)
```

Comment:
Please create one separate class attribute to store the singular form of the parent stream name.  The above things will not work if any new stream's plural ends with other than 's'.

Reply:
Yes, created separate attribute `parent_id_key` to store parent stream name.